### PR TITLE
Change language based on query parameter

### PR DESCRIPTION
**Pull request by veeraalt**
# Description

This PR adds a possibility of linking to a specific page in A+ in a specific language or viewing content with different language without having to change and save it in profile or in course page, by using language query parameter at the end of the url. 
For example, adding ?hl=en at the end of the url changes the content to English if it's supported for the page. If not, an error is raised.

* If the language is supported for A+ but not for this course, a 404 html page is shown with a list of available languages and links for each.
520-error1

* If the language is not supported for A+ at all, a 404 page is shown with a link to the material in default language.
404-default

Language parameter stays in the url when navigating through page, except when user changes language either on course page
or on profile page.

This is achieved by adding an event listener to `assets/js/aplus.js` that listens to click events and copies query parameters from the current url to the next when a link is clicked. When user changes language either on course page or on profile page, the language query parameter is removed from the url.

1/3 of issue https://github.com/apluslms/a-plus/issues/520

# Testing

* Existing unit tests pass

* Testing has been done manually by navigating through pages and checking that the language parameter changes the content language and also stays in the url, and that the errors are raised properly.

# Have you updated the README or other relevant documentation?

* No, I'm not sure if it's needed.

# Is it [Done](https://wiki.aalto.fi/display/EDIT/Definition+of+Done)?

- [ ] I (developer) have created unit tests and the tests pass
- [ ] I (developer) have created functional tests (Selenium tests) if applicable
- [x] I (developer) have tested the changes manually
- [x] Reviewer has finished the code review
- [x] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

### CODE REVIEW

File: course/viewbase.py [hunk 1]
```diff
@@ -117,6 +123,20 @@ def get_course_instance_object(self):
...
+        if isinstance(exc, TranslationNotFound):
+            instance_languages = self.instance.language
+            path_info = self.request.META.get('PATH_INFO')
+            languages = []
+            if instance_languages.startswith("|"):
+                languages = instance_languages.strip("|").split("|")
+            else:
+                languages = [instance_languages]
+            for i, lang in enumerate(languages):
+                languages[i] = {"name": get_language_info(lang)['name'], "url": path_info + '?hl=' + lang}
```
Comment (raphendyr) [inline course/viewbase.py:136]: You could use `update_url_params` from `lib/helpers.py` for joining the url.

File: course/viewbase.py [hunk 2]
```diff
@@ -117,6 +123,20 @@ def get_course_instance_object(self):
             course__url=self.kwargs[self.course_kw],
         )
 
+    def handle_exception(self, exc):
+        if isinstance(exc, TranslationNotFound):
+            instance_languages = self.instance.language
+            path_info = self.request.META.get('PATH_INFO')
```
Comment (raphendyr) [inline course/viewbase.py:129]: Maybe a bit nicer would be calling `request.get_full_path()`. Ouh, and `PATH_INFO` doesn't contain the full url when the app is bound to a sub path (see `SCRIPT_NAME` and `PATH_INFO` [here](https://wsgi.readthedocs.io/en/latest/definitions.html)).

File: templates/404.html [hunk 3]
```diff
@@ -16,12 +16,22 @@
                         <h1>404 Not found</h1>
                     </div>
                     {% include "_messages.html" %}
+                    {% if languages %}
+                    <div class="alert alert-danger">
+                        <p>The course is not available in this language.</p>
```
Comment (raphendyr) [inline templates/404.html:21]: These are missing a translation block.

File: lib/middleware.py [hunk 4]
```diff
@@ -1,19 +1,35 @@
...
 
 class LocaleMiddleware(MiddlewareMixin):
 
     response_redirect_class = HttpResponseRedirect
 
     def process_request(self, request):
-        if hasattr(request, 'user') and request.user.is_authenticated:
+        language = translation.get_language_from_request(request)
+        current_url = request.META.get('PATH_INFO')
+        last_url = request.META.get('HTTP_REFERER', '')
```
Comment (raphendyr) [inline lib/middleware.py:18]: Yeah, this doesn't look very nice way to do this. I'll return on this after thinking a bit of ideas where to look a better solution.

File: templates/404.html (hunk 3 above)
Comment (veeraalt) [reply templates/404.html:21]: For loop doesn't work inside the translation block because "'blocktrans' doesn't allow other block tags (seen 'for lang in languages') inside it". Is it a good solution to wrap only the text part inside translation block?

Comment (raphendyr) [reply templates/404.html:21]: You don't need to translate the for part, only the text before it

File: course/viewbase.py [hunk 5]
```diff
@@ -74,15 +76,20 @@ def get_resource_objects(self):
             # Try to find a language that is defined for this course instance
             # and apply it
             if self.instance.language:
-
+                instance_languages = self.instance.language.strip('|').split('|')
                 languages = []
+                url_lang = self.request.GET.get('hl')
+                if url_lang:
+                    if url_lang in instance_languages:
```
Comment (raphendyr) [inline course/viewbase.py:83]: This would benefit from the `instance_languages`, which is a set and not a list (i.e. you should have moved all 3 lines.

File: course/viewbase.py [hunk 6]
```diff
@@ -116,6 +123,15 @@ def get_course_instance_object(self):
             course__url=self.kwargs[self.course_kw],
         )
 
+    def handle_exception(self, exc):
+        if isinstance(exc, TranslationNotFound):
+            instance_languages = self.instance.language.strip("|").split("|")
+            url = remove_query_param_from_url(self.request.get_full_path(), 'hl')
+            for i, lang in enumerate(instance_languages):
+                instance_languages[i] = {"name": get_language_info(lang)['name'], "url": update_url_params(url, {'hl' : lang})}
```
Comment (raphendyr) [inline course/viewbase.py:131]: Hmm.. If A+ would be configured to use languages `en-fi` and `fi-fi` and the course would have `en` and `fi`, then this would generate links, which wouldn't use the correct value for the `hl` parameter. So, you probably need to iterate over `settings.LANGUAGES` and filter that, so that only those which have translation in the course are listed.

Comment (veeraalt) [reply course/viewbase.py:131]: Can you clarify why is the `settings.LANGUAGES` needed? If the course has languages `en` and `fi`, then the links would have `?hl=en` and `?hl=fi`, aren't those the right ones?

Comment (veeraalt) [reply course/viewbase.py:131]: Or should it be `?hl=en-fi` and `?hl=fi-fi`?

Comment (raphendyr) [reply course/viewbase.py:131]: Your other part of the code requires that the `?hl=<lang>` is found from the `settings.LANGUAGES`, thus `?hl=en` should result 404 if there is only `en-us` and `fi`.

Comment (veeraalt) [reply course/viewbase.py:131]: Maybe it's still quite unclear for me how the `en-fi` should work. If A+ has language `en-fi` and course is available in `en`, the course content should still surely be displayed in English? And I guess `hl` tag can only contain the first two characters of  the language code or how should it work with `en-us` for example?

Comment (raphendyr) [reply course/viewbase.py:131]: As the tanslation.activate is called with the value of the `hl`, then the `hl` should contain the full language info (i.e. locale = language + location). The second code, in course instance, then shortens this locale to just tlanguage, which are used to select content from the course.

In other words, all Django stuff expect the language parameter to be fully qualified name (e.g. `en-us`), so it can select correct translation file and formats. However, the course content is translated only with 2 letter codes, so we need to shorten the `en-us` to just `en` in all course material related places (and assessment tools, like grading).

Comment (veeraalt) [reply course/viewbase.py:131]: Thanks, that clears some things up but I'm somewhat confused about the formats still.

This is more a concern of https://github.com/apluslms/a-plus/pull/604, but that depends on how we solve the language locale/region issues here:

If we added the `formats/en-fi/formats.py` folder and it would be possible for the language to be `en`, `en-fi` and even `en-us` or `en-gb` all in the same session, would the date and time formats then vary between the ones defined in `formats/en/formats.py` and `formats/en-fi/formats.py`? That would be very confusing.

I don't see the point in supporting these different language formats (other than maybe for displaying dates and times) because it's not even possible for users to pick `en-fi`, `en-gb` or `en-us` as the preferred language, there is only `English`, so in which case `en-fi` would come up?

Comment (veeraalt) [reply course/viewbase.py:131]: Maybe it would be simpler if the language could be only `fi` or `en` and there was a separate `date_format` field in `UserProfile` that could be changed from profile page, but I'm not sure how displaying the correct format could be implemented.

Comment (raphendyr) [reply course/viewbase.py:131]: Let's write this down. Back in the day at times of UNIX, POSIX standard specified [locale syntax](https://en.wikipedia.org/wiki/Locale_(computer_software)) `[language[_territory][.codeset][@modifier]]`, e.g., `en_US.UTF-8` and `fi_FI.UTF-8` (the charset is now-a-days pretty useless as unicode contains it all, things were different before it). Another relevant info is how [HTTP specified](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language) `Accept-Language` header (e.g. `en`, `en-US`, `fi`) (see also [BCP 47 / IETF language tag](https://en.wikipedia.org/wiki/IETF_language_tag)). The problem which people has tried to solve is how to differentiate differences in the language.

A good question is why formats and languages are typically connected like this. Like why `fi_FI` defines formats used in Finalnd? Technically, that string contains all required information. The `en_FI` could be parsed as show texts with language English, but use geographical information related to Finland. For the latter, there is clear definition how numbers and such are represented in our location. Interestingly in Switzerland, the location part is not enough to define the used formats, thus both the language and the location are required to get the exact information.

Many distributions don't include all possible combinations for locales, e.g., `en_FI` doesn't exist in Debian or Ubuntu. However, it's part of the [Unicode Common Locale Data Repository](http://cldr.unicode.org/). This was the best reference to say that it's a valid locale string, which I could find.

So, languages _often_ requires the geographical location to represent the whole UI correctly (including formats).

However, we have so far went with the simpler language formats for course material. Though, I think we should start to plan that the course language code could include possibly include the location information too, at least to some extend. The full IETF standard would be very complicated though. So, for now we use only the two letter code to indicate the language for the material. As the locale to language map is always surjection (i.e. for every language, there can exist one or more locales). Thus, we can always map any locale to a language, but we can't map any language to one locale.

That said, Django should need to provide those formats files, as that information should be available via some library, but apparently that isn't the case yet. Thus, for the formats, I think we need to provide definitions for those locales, which we would like to use, but are not covered by the Django yet.

Next problem. The `hl` parameter (and userprofile.language and enrollment.language) requires the fully qualified locale name always, so the `translation.activate()` can operate correctly, as it needs to find the correct formats.py file too. If we think A+ as an application, which is used to read course material, then we can think that the application is always configured to show the UI with specific locale and not just with specific language. However, the application would select the best matching material from the database for that locale.

I guess, the biggest challenges are the cases when  the best information comes from `Accept-Language` and it doesn't include the location (e.g. just `en`). Maybe that is when we need to guess the best possible language, and provide way to change it.

Another annoyance would be that when viewing the material in English (US), we could provide button to change the language to English (UK), but as that would so the same content, but only change how the datetimes are represented, it could be very confusing to the user. I'm not sure what would be the best. Luckily, it's likely that there exist only one English definition in the settings, which would mean that typically locales and languages would have one to one mapping.

Maybe this contains some valuable information, maybe not. Have to say that things didn't become super clear for me either, even though I spend the evening reading about this stuff.

To summarise, the application (A+) should use the most accurate locale information (e.g. `en-us`, `en-fi`, `fi-fi`), when possible, and would select the best matching course material. If no version of the material is available for the locale, then A+ would show 404 with the list of possible versions (even if there would be multiple English variants).

File: lib/middleware.py [hunk 7]
```diff
@@ -1,19 +1,27 @@
...
+        query_language = request.GET.get('hl')
+        if query_language:
+            language = query_language
+            lang_codes = [lang[0][:2] for lang in settings.LANGUAGES]
+            if language and (language[:2] not in lang_codes):
+                raise Http404('Page is not available in this language.')   
+        elif hasattr(request, 'user') and request.user.is_authenticated:
             language = request.user.userprofile.language
-        else:
-            language = translation.get_language_from_request(request)
```
Comment (raphendyr) [inline lib/middleware.py:16]: This `else` branch could be restored as it would make more sense with the current code. Would remove the execution of `get_language_from_request` in case it comes from somewhere else.

File: course/viewbase.py [hunk 8]
```diff
@@ -74,20 +76,25 @@ def get_resource_objects(self):
...
+                instance_languages = set(instance_languages)
 
                 languages = []
+                query_language = self.request.GET.get('hl')
+                if query_language:
+                    if query_language[:2] in instance_languages:
+                        languages.append(query_language)
+                    else:
+                        raise TranslationNotFound
                 if self.user_course_data and self.user_course_data.language:
```
[the remark this answers is not in the document]
Comment (veeraalt) [reply course/viewbase.py:90]: Yes, the part where we look for enrollment and profile languages could be skipped, the last part should be still included because it checks that the course is available in the (query) language.

File: assets/js/aplus.js [hunk 9]
```diff
@@ -629,5 +629,26 @@ $(function() {
...
+            event.target.nodeName === 'A'
+            && event.target.href
+            && event.target.protocol === window.location.protocol
+            && event.target.host === window.location.host // hostname:port
+            && (event.target.search === '' || event.target.search === '?') // no search
+        ) {
+            event.preventDefault(); // stop default redirect
+            event.target.search = window.location.search; // copy query params
+            window.location.href = event.target.href; // redirect
+            event.target.hreflang = window.location.hreflang
```
[the remark this answers is not in the document]
Comment (veeraalt) [reply assets/js/aplus.js:649]: So the last line might be unnecessary?

File: course/viewbase.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (raphendyr) [reply course/viewbase.py:90]: Isn't that done already above? `if query_language[:2] in instance_languages:`

File: assets/js/aplus.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (raphendyr) [reply assets/js/aplus.js:649]: To me understanding, it has no effect (except removes hreflang if that exists).

File: course/viewbase.py (hunk 6 above)
Comment (veeraalt) [reply course/viewbase.py:131]: Thank you for the clarification, I still have a few questions about the logic/implementation:
1. Do we want to change the `LANGUAGES` in settings.py to contain only `('en-fi', 'English (FI)'), ('fi-fi', 'Finnish (FI)')` (and maybe also 'en-gb' and 'en-us') or just add those two and keep `('en', 'English'), ('fi', 'Finnish')` there too? If we replace the shorter formats completely in this PR, the `formats.py` files probably need to be replaced here also.
2. A course is only available in English, and user tries to access it in Finnish, what languages are there in the 404 error page, i.e. is there only `English` as the course language does not contain information about location or is there `English (FI)`, `English (UK)` etc?
3. When using only `?hl=en`, do we want to show the error page or direct the user to `?hl=en-fi`? I think it would be better to not show the error but the logic might get complicated if we need to guess the longer language code from the shorter one and change it every time. 
4. How about with Finnish? When we probably have only `fi-fi`, it would make sense to accept both `fi` and `fi-fi` as language codes and not show the error page when query parameter is `?hl=fi`.

There's so many unclear things about this that we need to find some compromise solution in scope of this PR if we want to get this somehow done and merged.

File: course/viewbase.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (veeraalt) [reply course/viewbase.py:90]: Oh that is true, don't know how I didn't notice that.

File: course/viewbase.py (hunk 6 above)
Comment (veeraalt) [reply course/viewbase.py:131]: I think that `('en', 'English'), ('fi', 'Finnish')` has to be there as long as course code can be only 2 characters. Then we might be left with this kind of error message:
404-english

Comment (raphendyr) [reply course/viewbase.py:131]: > Do we want to change the LANGUAGES in settings.py to contain only ('en-fi', 'English (FI)'), ('fi-fi', 'Finnish (FI)') (and maybe also 'en-gb' and 'en-us') or just add those two and keep ('en', 'English'), ('fi', 'Finnish') there too? If we replace the shorter formats completely in this PR, the formats.py files probably need to be replaced here also.

To help with #604, I would say we want to have `en-fi` and `fi` (`fi-fi` doesn't add any more information).

> A course is only available in English, and user tries to access it in Finnish, what languages are there in the 404 error page, i.e. is there only English as the course language does not contain information about location or is there English (FI), English (UK) etc?

They should see list of all possible languages, which would match, i.e., `English (US), English (FI), Finnish` for a course with langs `['en', 'fi']` and LANGUAGES `['en-fi', 'en-us', 'fi']`, when they try to access the course with some other language. This isn't perfect, but I think the best we could do.

Better option would be having a different UI language and course language, but that is not the case for current design. However, I think in most of the cases, the system will be configured with a single English locale.

> When using only ?hl=en, do we want to show the error page or direct the user to ?hl=en-fi? I think it would be better to not show the error but the logic might get complicated if we need to guess the longer language code from the shorter one and change it every time.

Technically, we should show an error, but as you said that it would be annoying. Hence, let's make an educated guess by redirecting the user to the first matching locale in the LANGUAGES list (it's an ordered list).

> How about with Finnish? When we probably have only fi-fi, it would make sense to accept both fi and fi-fi as language codes and not show the error page when query parameter is ?hl=fi.

I think it should be sufficient to use just `fi` in the defined languages (it's fully qualified name, as there is no sub languages. Related languages have their own tags, e.g., `fit` for [Tornedalen Finnish](https://en.wikipedia.org/wiki/Me%C3%A4nkieli)).

Thus, we probably should also do lookup for a shorter `hl` entry too. I.e., if we have `?hl=tag1-tag2-tag3-Tag4`, we would first check if there is `tag1-tag2-tag3-Tag4` in LANGUAGES, if not then check `tag1-tag2-tag3`, `tag1-tag2` and finally `tag1`. And as always, select the first from the list of matches and redirect.

As a last remark, I would expect the settings to contain something like this:

```python
LANGUAGES = (
  ('en-FI', _('English')),
  ('fi', _('Finnish')),
  ('sv', _('Swedish')),
)
```

Or in the US:

```python
LANGUAGES = (
  ('en-US', _('English')),
)
```

Or in the Mexico (I guess):

```python
LANGUAGES = (
  ('es-MX', _('Spanish')),
  ('en-US', _('English')),
)
```

p.s. the region code should be styled with uppercase letters, but I'm not sure how Django handles that in this settings value. The `translation.activate` handles it correctly, if I remember right.

These details (i.e. `es-MX` vs. `es`) only help with formats in our case for now.

Comment (veeraalt) [reply course/viewbase.py:131]: Your suggestion seems otherwise good, but I'm a bit worried whether things are going to break if we remove `en` from `settings.py` when it's still used as course instance language. I tested that today with the test database and it deleted the `en` from language field in course instances where it was used, I suppose because it wasn't supported in `settings.py`.
We could have `English (US)` or `English (UK)` with language code `en` as the default, and option to change it to `English (FI)` at least from profile page, which then allows to use the Finnish formats.
What do you think of this? 
```
LANGUAGES = [
    ('en-FI', 'English (FI)'),
    ('en', 'English (US)'),
    ('fi', 'Finnish'),
]
```

Comment (Mankro) [reply course/viewbase.py:131]: I think we have to have a discussion in a meeting about these language code issues (on Tuesday?). I'm afraid @raphendyr is making this really complicated and I hope it wouldn't need to be so. I can see this is not "as easy as it looks" and that languages and localizations have many twists. However, I would like to consider whether we could still aim for a simple, minimal approach. That is, language choice would focus on the language, not the region, and short 2-character language codes would be used as much as possible. How about continuing with this after the daily?

Comment (veeraalt) [reply course/viewbase.py:131]: @Mankro Sounds good for me.

Comment (raphendyr) [reply course/viewbase.py:131]: Yes, rest of the software doesn't yet handle language codes longer than 2 characters. I would like to get this one part working, so rest of the components can be fixed while moving forward. In the end, there are only few places which need to be fixed.

> However, I would like to consider whether we could still aim for a simple, minimal approach. That is, language choice would focus on the language, not the region, and short 2-character language codes would be used as much as possible.

I wish it was so simple. We are just mixing locale information and course languages to the same variable, where the course language doesn't often include enough information for the rest of the system.

Even ISO 639-3 codes (3 letter language codes) would already work for many cases, but not all, e.g., `pt-BR` (though, I have no idea why it doesn't have ISO code).

Comment (raphendyr) [reply course/viewbase.py:133]: We made a decision to limit the language strings to 2 character codes (i.e. ISO 639-2 codes), which should work for the currently known use cases and has less change to break stuff. The code should in most cases ensure the longer or invalid codes are cut or dropped, so that only valid codes are used.

File: lib/middleware.py [hunk 10]
```diff
@@ -1,16 +1,25 @@
...
 
     def process_request(self, request):
-        if hasattr(request, 'user') and request.user.is_authenticated:
+        query_language = request.GET.get('hl')
+        if query_language:
+            language = query_language
+            lang_codes = [lang[0][:2] for lang in settings.LANGUAGES]
+            if language and (language[:2] not in lang_codes):
+                raise Http404('Page is not available in this language.')   
+        elif hasattr(request, 'user') and request.user.is_authenticated:
```
Comment (raphendyr) [inline lib/middleware.py:22]: I think this requires the is_anonymous check too.

File: lib/middleware.py [hunk 11]
```diff
@@ -1,16 +1,25 @@
...
 class LocaleMiddleware(MiddlewareMixin):
 
     response_redirect_class = HttpResponseRedirect
 
     def process_request(self, request):
-        if hasattr(request, 'user') and request.user.is_authenticated:
+        query_language = request.GET.get('hl')
+        if query_language:
+            language = query_language
+            lang_codes = [lang[0][:2] for lang in settings.LANGUAGES]
```
Comment (raphendyr) [inline lib/middleware.py:19]: A set could be used here, so the `in` check below would be faster.

File: templates/404.html [hunk 12]
```diff
@@ -16,12 +16,24 @@
                         <h1>404 Not found</h1>
                     </div>
                     {% include "_messages.html" %}
+                    {% if languages %}
+                    <div class="alert alert-danger">
+                        {% blocktrans %}
```
Comment (raphendyr) [inline templates/404.html:21]: Can you add `trimmed` here, so the translation file wouldn't include `\n` chars.

**Review by raphendyr — APPROVED** [review]
It's good from my point of view. I think you could squash and rebase and stuff it now.

**Review by raphendyr — APPROVED** [review]

File: course/viewbase.py [hunk 13]
```diff
@@ -117,6 +126,15 @@ def get_course_instance_object(self):
             course__url=self.kwargs[self.course_kw],
         )
 
+    def handle_exception(self, exc):
+        if isinstance(exc, TranslationNotFound):
```
Comment (Mankro) [inline course/viewbase.py:130]: I tried to test this myself. When I used a non-existing language code in the URL of a course page (`?hl=sv`), the 404 page did not show the available languages. The exception in this function was apparently not of type `TranslationNotFound` and the `languages` template context variable was then not defined. The exception was actually `No CourseInstance matches the given query. <class 'django.http.response.Http404'>` (from a debug print). Maybe that comes from the `LocaleMiddleware` instead of `CourseInstanceBaseMixin`. Can you try this yourself and fix it if the bug is real?

**Review by Mankro — CHANGES_REQUESTED** [review]
This looks good, but I found one bug.

File: course/viewbase.py (hunk 13 above)
Comment (Mankro) [reply course/viewbase.py:130]: I just noticed one important point: `sv` is not enabled in the A+ settings.py. If I use `fi` in an English course, then the 404 page shows the link to the English version of the course page. (`fi` and `en` are the default languages in settings.py.)

Comment (raphendyr) [reply course/viewbase.py:130]: I think this was a undocumented design decisionI think this was a undocumented design decision.

The `?hl=` is first validated in the middleware, where it has to be something defined in the settings file, hence the `?hl=sv` raises error there. If the variable is valid, then the processing continues to view classes, which include this better solution.

Making the course available to the middleware wouldn't be feasible and ignoring the language error in the middleware might raise other problems.

In short, 404 without language list is application level error, while the other is course level error. Though, that application page _could_ list all configured languages instead of not mentioning anything.

Comment (Mankro) [reply course/viewbase.py:130]: I see. Which one is better in the following? If `LocaleMiddleware` raises 404 due to the GET query `hl` parameter (unknown language for the site), then...

* the 404 page shows a link to the same page without any `hl` query parameter. That link would work if the `hl` parameter was the only reason for the 404 error. The user would probably get at least somewhere and understand that the language is not supported. Better to reach some content than to assume that the whole course URL is broken.
* the 404 page shows language links (using `hl` query parameter) to all language versions of the page. All languages would be taken from the site-wide settings.py variable. Not all courses support those languages, so the links would lead to a new 404 page that would finally show working links to the course page.

Comment (Mankro) [reply course/viewbase.py:130]: The first alternative without the `hl` seems more reasonable after our daily discussion. Would the link text be something like "open this page in the default language"?

Comment (raphendyr) [reply course/viewbase.py:130]: One alternative would be doing an automatic redirect, but add like `?msg=automated redirect due to invalid hl parameter`, which would be rendered as a warning flash. I don't remember if we had such a feature yet or not.

Comment (Mankro) [reply course/viewbase.py:130]: Since time is running out, we can only either add a link to the 404 page or keep it as is without any link.

Comment (raphendyr) [reply course/viewbase.py:130]: I personally don't see a huge problem, as it should be less likely to include values which are not valid for any course than have a wrong value for that course (or that specific page in a course).

Hence, I would go what there is now and make issue to fix this better later.

Comment (veeraalt) [reply course/viewbase.py:130]: I just managed to do the changes regarding the first approach, so I could push those and let me know if it looks good enough or do we want to discard them and go by what we have now.

Comment (raphendyr) [reply course/viewbase.py:130]: That works too. Btw. do those pages still return 404?

Comment (raphendyr) [reply course/viewbase.py:130]: > Btw. do those pages still return 404?

They do, so that is good :)

**Review by Mankro — APPROVED** [review]
Great, the 404 got the small improvement that we discussed. I think this could be squashed and merged. @raphendyr any objections?

File: lib/middleware.py [hunk 14]
```diff
@@ -1,16 +1,28 @@
+from urllib.parse import urlparse
+
 from django.conf import settings
-from django.http import HttpResponseRedirect
+from django.http import Http404, HttpResponseRedirect
 from django.utils import translation
 from django.utils.cache import patch_vary_headers
 from django.utils.deprecation import MiddlewareMixin
+from django.shortcuts import redirect, render_to_response
```
Comment (raphendyr) [inline lib/middleware.py:8]: `redirect` is not used?

File: lib/middleware.py [hunk 15]
```diff
@@ -1,16 +1,28 @@
+from urllib.parse import urlparse
+
 from django.conf import settings
-from django.http import HttpResponseRedirect
+from django.http import Http404, HttpResponseRedirect
```
Comment (raphendyr) [inline lib/middleware.py:4]: `Http404` is not used?

File: lib/middleware.py (hunk 14 above)
Comment (raphendyr) [inline lib/middleware.py:8]: redirect is not used?

File: lib/middleware.py [hunk 16]
```diff
@@ -1,16 +1,28 @@
+from urllib.parse import urlparse
```
Comment (raphendyr) [inline lib/middleware.py:1]: `urlparse` is not used?

**Review by raphendyr — COMMENTED** [review]
Found some small things. You can fix and rebase those. Thanks.
